bftengine/src/preprocessor/PreProcessor.cpp

```diff
@@ -645,18 +727,25 @@ void PreProcessor::handleClientPreProcessRequestByNonPrimary(ClientPreProcessReq
   const auto msgSize = clientReqMsg->size();
   const auto cid = clientReqMsg->getCid();
   // Register a client request message with an empty PreProcessRequestMsg to allow follow up.
-  if (registerRequest(move(clientReqMsg), PreProcessRequestMsgSharedPtr())) {
+  if (registerRequest(move(clientReqMsg), PreProcessRequestMsgSharedPtr(), reqOffsetInBatch)) {
     LOG_INFO(logger(),
              "Start request processing by a non-primary replica"
-                 << KVLOG(reqSeqNum, cid, clientId, senderId, reqTimeoutMilli));
+                 << KVLOG(reqSeqNum, cid, clientId, reqOffsetInBatch, senderId, reqTimeoutMilli));
+    if (arrivedInBatch) return;  // Need to re-send the whole batch to the primary
     sendMsg(msgBody, myReplica_.currentPrimary(), msgType, msgSize);
     LOG_DEBUG(logger(),
-              "Sent ClientPreProcessRequestMsg" << KVLOG(reqSeqNum, cid, clientId) << " to the current primary");
+              "Sent ClientPreProcessRequestMsg" << KVLOG(reqSeqNum, cid, clientId, reqOffsetInBatch)
+                                                << " to the current primary");
   }
 }
 
-const char *PreProcessor::getPreProcessResultBuffer(uint16_t clientId) const {
-  return preProcessResultBuffers_[getClientReplyBufferId(clientId)].data();
+const char *PreProcessor::getPreProcessResultBuffer(uint16_t clientId, uint16_t reqOffsetInBatch) const {
+  const auto bufferOffset = clientId - numOfReplicas_ + reqOffsetInBatch;
```

Comment:
Maybe a comment would help clarify this formula?

Reply:
Will add.